## 计算方法Exp3

#### 06219109 孙寒石

### 用龙贝格方法计算：
$$\int_0^{\frac{\pi}{2}}\sin(2\cos x)\sin^2 xdx$$

In [1]:
import math

In [2]:
def f(x):
    return math.sin(2*math.cos(x))*math.sin(x)*math.sin(x)

In [3]:
def Romberg(a,b,f,eps):
    n = 1
    h = b - a
    T1 = (b-a)/2*(f(a)+f(b))
    while(1):
        temp = 0
        for i in range(n):
            x=a+i*h+h/2
            temp=temp+f(x)
        T2=(T1+temp*h)/2
        if(abs(T2-T1)<eps):
            return T2
        S2=T2+(T2-T1)/3
        if(n==1):
            T1=T2
            S1=S2
            h=h/2
            n=n*2
            continue
        C2=S2+(S2-S1)/15
        if(n==2):
            C1=C2
            T1=T2
            S1=S2
            h=h/2
            n=n*2
            continue
        R2=C2+(C2-C1)/63
        if (n==4):
            R1=R2
            C1=C2
            T1=T2
            S1=S2
            h=h/2
            n=n*2
            continue
        if(abs(R2-R1)<eps):
            return R2
        R1=R2
        C1=C2
        T1=T2
        S1=S2
        h=h/2
        n=n*2
        

In [4]:
pi = 3.141592653589793
eps = 5e-6
print(Romberg(0,pi/2,f,eps))

0.5079670457310737


#### 所以最后的结果为 $$\int_0^{\frac{\pi}{2}}\sin(2\cos x)\sin^2 xdx=0.5079670457310737$$